### PR TITLE
(towards #2132) LFRic integration tests

### PR DESCRIPTION
**Pull request by sergisiso**

### CODE REVIEW

File: config/psyclone.cfg [hunk 1]
```diff
@@ -63,7 +63,7 @@ access_mapping = gh_read: read, gh_write: write, gh_readwrite: readwrite,
 # that it iterates over dofs. This is currently only the case for
 # builtins. If annexed dofs are computed then in certain cases we
 # remove the need for a halo exchange call.
-COMPUTE_ANNEXED_DOFS = false
+COMPUTE_ANNEXED_DOFS = true
```
Comment (sergisiso) [inline config/psyclone.cfg:66]: @rupertford @arporter This is needed for properly running LFRic. Should we change this default value? Or do you prefer that I sed replace it in the github workflow at let it be false by default.

Same with NUM_ANY_DISCONTINUOUS_SPACE below, LFRic uses 20.

**Comment by sergisiso** [comment]
@arporter and @rupertford This is ready for a first review, it:
- Adds the LFRic CI tests.
- Has scripts in examples/lfric/scripts for each main LFRic transformation, both in isolation and all combined.
- Has the necessary bugfixes to make the mentioned scripts work.

File: src/psyclone/psyir/transformations/intrinsics/matmul2code_trans.py [hunk 2]
```diff
@@ -454,18 +471,19 @@ def _apply_matrix_matrix(node):
         assign = Assignment.create(result_ref.copy(), rhs)
         # Create ii loop and add the above code as a child
         # Work out the bounds
-        lower_bound, upper_bound, step = _get_array_bound(matrix1, 0)
+        lower_bound, upper_bound, step = _get_array_bound(matrix1, 1)
```
Comment (arporter) [inline src/psyclone/psyir/transformations/intrinsics/matmul2code_trans.py:474]: Oops! Embarrassing :-( Are these changes needed to make LFRic give the same answers? (i.e. LFRic was broken and this fixes it?)

**Review by arporter — CHANGES_REQUESTED** [review]
Thanks Sergi. Just some things to tidy up.
Embarrassing that matrix-matrix wasn't right! (I could have sworn I tested it by compiling and running some code but there we go.)
Ref. guide builds without new warnings.
I've set the integration tests going and will check on them next time.

File: examples/lfric/scripts/everything_everywhere_all_at_once.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+                                     OMPParallelTrans, \
+                                     Dynamo0p3RedundantComputationTrans, \
+                                     Dynamo0p3AsyncHaloExchangeTrans, \
+                                     MoveTrans, \
+                                     TransformationError
+
+ENABLE_REDUNDANT_COMPUTATION = True
+ENABLE_ASYNC_HALOS = True
+ENABLE_OMP_COLOURING = True
+ENABLE_LOOP_FUSING = False
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply examples/lfric/scripts/everything_everywhere_all_at_once.py:59]: Removed one and added a TODO comment for the other.

File: examples/lfric/scripts/everything_everywhere_all_at_once.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+                    mtrans.apply(ahex, schedule.children[location_cursor])
+                    location_cursor += 1
+                except TransformationError:
+                    pass
+
+
+        if ENABLE_LOOP_FUSING:
+            # Fuse Loops when possible
+            idx = len(schedule.children) - 1
+            while idx > 0:
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply examples/lfric/scripts/everything_everywhere_all_at_once.py:112]: Ups. Done now

File: examples/lfric/scripts/everything_everywhere_all_at_once.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sergisiso) [reply examples/lfric/scripts/everything_everywhere_all_at_once.py:59]: Finally I removed both, I will look at these in the future

File: examples/lfric/scripts/everything_everywhere_all_at_once.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+                node = schedule.children[idx]
+                prev_node = schedule.children[idx-1]
+                try:
+                    lf_trans.apply(prev_node, node, {"same_space": True})
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply examples/lfric/scripts/everything_everywhere_all_at_once.py:116]: Removed the whole transformation for now as it was disabled anyway

File: examples/lfric/scripts/everything_everywhere_all_at_once.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+
+        if ENABLE_KERNEL_CONSTANTS or ENABLE_INTRINSIC_INLINING:
+            # Inline kernels when possible
+            for kernel in schedule.coded_kernels():
+                try:
+                    inline_trans.apply(kernel)
+                except TransformationError:
+                    pass
+
+    # Then transform all the kernels inlined into the module
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply examples/lfric/scripts/everything_everywhere_all_at_once.py:143]: Some of them can be, like async transformation and separate the start communication. But others are just finding the node we are interested in and try/expect the transformations. I think pylint limits are very strict in this case.

File: examples/lfric/scripts/async_halo_exchanges.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+from psyclone.dynamo0p3 import DynHaloExchange, DynHaloExchangeStart
+from psyclone.transformations import Dynamo0p3AsyncHaloExchangeTrans, \
+    MoveTrans, TransformationError
+
+
+def trans(psy):
+    '''A sample transformation script to demonstrate the use of asynchronous
+    halo exchanges with overlapping compute and communication for the
+    most costly halo exchanges in the (current version of the) LFRic model.
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply examples/lfric/scripts/async_halo_exchanges.py:51]: It is general purpose, I have applied it to the whole LFRic successfully, I just forgot to update this comment.

The eg11 is just specific to a certain algorithm and can not be applied generically.

File: examples/lfric/scripts/kernel_print.py [hunk 8]
```diff
@@ -40,20 +40,29 @@
 '''
 from __future__ import print_function
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply examples/lfric/scripts/kernel_print.py:41]: Done

File: examples/lfric/scripts/kernel_print.py [hunk 9]
```diff
@@ -40,20 +40,29 @@
...
-            kernel_schedule = kernel.get_kernel_schedule()
-            kern = fortran_writer(kernel_schedule)
-            print(kern)
+            try:
+                kernel_schedule = kernel.get_kernel_schedule()
+                if kernel_schedule not in already_printed:
+                    kern = fortran_writer(kernel_schedule)
+                    print(kern)
+                    already_printed.append(kernel_schedule)
+            except Exception as err:
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply examples/lfric/scripts/kernel_print.py:64]: Done

File: src/psyclone/tests/psyir/transformations/intrinsics/matmul2code_trans_test.py [hunk 10]
```diff
@@ -572,6 +570,66 @@ def test_validate14():
     trans.validate(matmul)
 
 
+def test_validate_matmat_with_slices_on_rhs(fortran_reader):
+    '''
+    Check the apply method works when there are already symbols present
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/psyir/transformations/intrinsics/matmul2code_trans_test.py:575]: Ups. Updated now.

File: src/psyclone/tests/psyir/transformations/intrinsics/matmul2code_trans_test.py [hunk 11]
```diff
@@ -572,6 +570,66 @@ def test_validate14():
...
+        "subroutine my_sub()\n"
+        "  real, dimension(2,6) :: jac\n"
+        "  real, dimension(6,3) :: jac_inv\n"
+        "  real, dimension(10,10) :: result\n"
+        "  result(2:4,2:5) = matmul(jac(:,:), jac_inv(:,:))\n"
+        "end subroutine my_sub\n")
+    trans = Matmul2CodeTrans()
+    assign = psyir.walk(Assignment)[0]
+    with pytest.raises(TransformationError) as excinfo:
+        trans.apply(assign.rhs)
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/psyir/transformations/intrinsics/matmul2code_trans_test.py:589]: Done

File: src/psyclone/tests/psyir/transformations/intrinsics/matmul2code_trans_test.py [hunk 12]
```diff
@@ -572,6 +570,66 @@ def test_validate14():
...
+    assert ("To use matmul2code_trans on matmul, each range on the result "
+            "variable 'result' must be a full range but found "
+            "result(2:4,2:5)" in str(excinfo.value))
+
+
+def test_validate_matmat_with_same_mem(fortran_reader):
+    '''
+    Check the apply method works when there are already symbols present
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/psyir/transformations/intrinsics/matmul2code_trans_test.py:597]: Done

File: src/psyclone/tests/psyir/transformations/intrinsics/matmul2code_trans_test.py [hunk 13]
```diff
@@ -572,6 +570,66 @@ def test_validate14():
...
+    psyir = fortran_reader.psyir_from_source(
+        "subroutine my_sub()\n"
+        "  real, dimension(2,2) :: jac\n"
+        "  real, dimension(2,2) :: jac_inv\n"
+        "  jac = matmul(jac(:,:), jac_inv(:,:))\n"
+        "end subroutine my_sub\n")
+    trans = Matmul2CodeTrans()
+    assign = psyir.walk(Assignment)[0]
+    with pytest.raises(TransformationError) as excinfo:
+        trans.apply(assign.rhs)
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/psyir/transformations/intrinsics/matmul2code_trans_test.py:610]: Done

File: src/psyclone/tests/psyir/transformations/intrinsics/matmul2code_trans_test.py [hunk 14]
```diff
@@ -572,6 +570,66 @@ def test_validate14():
...
+        "  real, dimension(2,2) :: jac\n"
+        "  real, dimension(2,2), pointer :: jac_inv\n"
+        "  real, dimension(2,2), pointer :: result\n"
+        "  result = matmul(jac(:,:), jac_inv(:,:))\n"
+        "end subroutine my_sub\n")
+    trans = Matmul2CodeTrans()
+    assign = psyir.walk(Assignment)[0]
+    with pytest.raises(TransformationError) as excinfo:
+        trans.apply(assign.rhs)
+    print(excinfo.value)
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/psyir/transformations/intrinsics/matmul2code_trans_test.py:626]: Done

File: src/psyclone/tests/psyir/transformations/intrinsics/matmul2code_trans_test.py [hunk 15]
```diff
@@ -689,45 +747,6 @@ def test_apply_matvect_no_indices(tmpdir, fortran_writer):
...
-    assert (
-        "subroutine my_kern()\n"
-        "  integer, parameter :: idx = 3\n"
-        "  real, dimension(5,10,15) :: x\n"
-        "  real, dimension(10,20,10) :: y\n"
-        "  real, dimension(10) :: result\n"
-        "  integer :: i\n"
-        "  integer :: j\n"
-        "\n"
-        "  do i = 1, 5, 1\n"
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/psyir/transformations/intrinsics/matmul2code_trans_test.py:720]: Yes

File: src/psyclone/psyir/transformations/intrinsics/matmul2code_trans.py [hunk 16]
```diff
@@ -291,28 +291,45 @@ def validate(self, node, options=None):
...
+        if result.children:
+            for idx, child in enumerate(result.children):
+                if isinstance(child, Range) and not result.is_full_range(idx):
+                    raise TransformationError(
+                        f"To use matmul2code_trans on matmul, each range on "
+                        f"the result variable '{result.name}' must be a full "
+                        f"range but found {result.debug_string()}")
+
+        # Make sure the result is not one of the MATMUL operands
+        for ref in result.walk(Reference):
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/transformations/intrinsics/matmul2code_trans.py:326]: You are right, the walk is not necessary.

File: src/psyclone/psyir/transformations/intrinsics/matmul2code_trans.py (hunk 2 above)
Comment (sergisiso) [reply src/psyclone/psyir/transformations/intrinsics/matmul2code_trans.py:474]: Yes, without this (and preventing the substitution of cases with same lhs and rhs operand) LFric does not converge in some sub-stepping and fails after reaching a iteration threshold.

File: examples/lfric/scripts/everything_everywhere_all_at_once.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (sergisiso) [reply examples/lfric/scripts/everything_everywhere_all_at_once.py]: but ended up removing it...

File: examples/lfric/scripts/kernel_print.py [hunk 17]
```diff
@@ -40,20 +40,29 @@
 '''
 from __future__ import print_function
 from psyclone.psyir.backend.fortran import FortranWriter
+from psyclone.transformations import TransformationError
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply examples/lfric/scripts/kernel_print.py]: deleted

**Comment by sergisiso** [comment]
@arporter This is ready for another review. I cut down the couple of problematic transformations from the everything script for now as I want to explore those with more time and are not used currently in LFRic anyway.

I also deployed the last commit here to the GH runner, so you may find it already completed when reviewing.

File: examples/lfric/scripts/everything_everywhere_all_at_once.py [hunk 18]
```diff
@@ -0,0 +1,156 @@
...
+                        not in const.VALID_DISCONTINUOUS_NAMES:
+                    ctrans.apply(loop)
+
+            # Add OpenMP to loops unless they are over colours or are null
+            for loop in schedule.loops():
+                if loop.loop_type not in ["colours", "null"]:
+                    oregtrans.apply(loop)
+                    otrans.apply(loop, options={"reprod": True})
+
+        if ENABLE_KERNEL_CONSTANTS or ENABLE_INTRINSIC_INLINING:
```
[the remark this answers is not in the document]
Comment (arporter) [reply examples/lfric/scripts/everything_everywhere_all_at_once.py]: Ah, I see!

**Review by arporter — APPROVED** [review]
All requested changes have been made or successfully refuted.
Integration tests have run and are green.
Coverage is all fine.
Will proceed to merge.
